x-pack/plugins/cloud_security_posture/server/routes/status/status.ts

```diff
@@ -0,0 +1,174 @@
+import type { CspSetupStatus, Status } from '../../../common/types';
+import {
+  addRunningAgentToAgentPolicy,
+  getCspAgentPolicies,
+  getCspPackagePolicies,
+} from '../../lib/fleet_util';
+
+export const INDEX_TIMEOUT_IN_MINUTES = 10;
+
+const isFindingsIndexExists = async (esClient: ElasticsearchClient): Promise<boolean> => {
+  try {
+    const queryResult = await esClient.search({
+      index: LATEST_FINDINGS_INDEX_DEFAULT_NS,
+      query: {
+        match_all: {},
+      },
+      size: 1,
+    });
+
+    return !!queryResult.hits.hits.length;
+  } catch (e) {
+    return false;
+  }
+};
+
+const getHealthyAgents = (enrichedAgentPolicies: GetAgentPoliciesResponseItem[]): number =>
+  enrichedAgentPolicies.reduce(
+    (previousValue, currentValue) => previousValue + (currentValue.agents || 0),
+    0
+  );
+
+const getMinutesPassedSinceMoment = (momentInput: MomentInput): number =>
+  moment().diff(moment(momentInput), 'minutes');
+
+const getStatus = (
+  findingsIndexExists: boolean,
+  installedIntegrations: number,
+  healthyAgents: number,
+  minutesPassedSinceInstallation: number
+): Status => {
+  if (findingsIndexExists) return 'indexed';
+  if (installedIntegrations === 0) return 'not-installed';
+  if (healthyAgents === 0) return 'not-deployed';
+  if (minutesPassedSinceInstallation <= INDEX_TIMEOUT_IN_MINUTES) return 'indexing';
+  if (minutesPassedSinceInstallation > INDEX_TIMEOUT_IN_MINUTES) return 'index-timeout';
+
+  throw new Error('could not determine csp setup status');
+};
+
+const getCspSetupStatus = async (
+  esClient: ElasticsearchClient,
+  soClient: SavedObjectsClientContract,
+  packageService: PackageService,
+  packagePolicyService: PackagePolicyServiceInterface,
+  agentPolicyService: AgentPolicyServiceInterface,
+  agentService: AgentService
+): Promise<CspSetupStatus> => {
+  const [findingsIndexExists, installationPackageInfo, latestPackageInfo, installedIntegrations] =
+    await Promise.all([
+      isFindingsIndexExists(esClient),
+      packageService.asInternalUser.getInstallation(CLOUD_SECURITY_POSTURE_PACKAGE_NAME),
+      packageService.asInternalUser.fetchFindLatestPackage(CLOUD_SECURITY_POSTURE_PACKAGE_NAME),
+      getCspPackagePolicies(soClient, packagePolicyService, CLOUD_SECURITY_POSTURE_PACKAGE_NAME, {
+        per_page: 10000,
+      }),
+    ]);
+
+  const agentPolicies = await getCspAgentPolicies(
+    soClient,
+    installedIntegrations.items,
+    agentPolicyService
+  );
+
+  const enrichedAgentPolicies = await addRunningAgentToAgentPolicy(
+    agentService,
+    agentPolicies || []
+  );
+  const healthyAgents = getHealthyAgents(enrichedAgentPolicies);
+  const installedIntegrationsTotal = installedIntegrations.total;
+  const latestPackageVersion = latestPackageInfo.version;
+
+  const status = getStatus(
+    findingsIndexExists,
+    installedIntegrationsTotal,
+    healthyAgents,
+    getMinutesPassedSinceMoment(installationPackageInfo?.install_started_at || 0)
```

Comment:
while i understand your point, destructering from a potentially undefined object is problematic and will require another check to ensure this object exists before we destructure from it.
also, assigning `0` as the default value for `install_start_at` is less optimal in my eyes, since `install_start_at` is a string or undefined, but giving it a default value of 0 is adding another type option for a number.

right now it reads like "get minutes passed since date || number", I think it's overall less convoluted